### ISSUE TITLE
util/hlc: investigate different heuristic for incrementing logical ticks

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#18697]
Currently, `hlc.Clock.Now()` bumps the logical component of the clock whenever the physical clock reports a time that is less than the most recent time we've seen in a call to `hlc.Clock.Update()`. There is a special case for timestamp encoding in MVCC keys to elide the logical component if it is zero. So it would be useful to try and avoid non-zero logical components.

An alternative would be to increment `hlc.mu.timestamp.Walltime`. The downside to doing that is repeated calls to `Now()` can move `Walltime` arbitrarily far into the future. It also has the downside of making it less obvious when a timestamp has been logically incremented.

A middle-ground would be to allow a fixed number of `Walltime` ticks before incrementing the logical component. For example, we could allow 10000 `Walltime` ticks which corresponds to 10us. 

You might ask, how often does incrementing of the logical component occur? Quite frequently. On a 6-node cluster, one node incremented its logical component 35k times (~6.5% of the calls to `hlc.Clock.Now()`) during a 30s period. This isn't just a clock that is perpetually behind as the maximum logical clock value seen during that time was 41. 

Cc @tschottdorf, @bdarnell

**Comment by bdarnell** [comment]
> A middle-ground would be to allow a fixed number of Walltime ticks before incrementing the logical component. For example, we could allow 10000 Walltime ticks which corresponds to 10us.

What does this mean concretely? Increment `Walltime` by 1ns when it is less than `physicalNow() + 10µs`, and increment the logical component when it is greater? I think you'd need to allow much more deviation from the physical clock to make much difference. I think that even though we don't get pinned at one timestamp for a large number of logical ticks, we're still seeing these ticks for millisecond-scale clock variance across the nodes, instead of sub-microsecond scales from requesting timestamps faster than the resolution of the physical clock. It would be easy to test this empirically to see if this is right or not.

Some math: there were 538k calls to hlc.Clock.Now() in this 30s period, or 18k/sec. Assuming a uniform distribution (which is surely wrong), the time that we reached 41 logical ticks represented a 2ms clock deviation. So incrementing Walltime for up to 2ms would let us get rid of most of these logical tick increments. But is it worth it to save 4 bytes (pre-compression) on ~6.5% of keys? Messing with time seems like a risky way to accomplish this.

**Comment by petermattis** [comment]
> What does this mean concretely? Increment Walltime by 1ns when it is less than physicalNow() + 10µs, and increment the logical component when it is greater?

I was imagining something slight different. When we determine that `physicalNow() <= c.mu.timestamp.WallTime`, we increment `WallTime` for some number of "ticks". This would require an additional bit of state in `hlc` to remember when we started adding logical ticks to `WallTime`. The amount of physical time we keep doing this for is dependent on how fast we're calling `Now()`. It might be microseconds or milliseconds. 

> But is it worth it to save 4 bytes (pre-compression) on ~6.5% of keys? Messing with time seems like a risky way to accomplish this.

I love that statement because it isn't many jobs that get to "mess with time". Is it worth it? I'm not sure, this is just something I stumbled across and wanted to file an issue to investigate it. I've been assuming that `Logical` is always 0, but clearly that isn't true.

**Comment by tschottdorf** [comment]
I'd rather avoid this kind of optimization unless it's proven significant and we've exhausted higher-level options for avoiding these ticks. We should understand better where these ticks are introduced (in a cluster with clock offset << latency). Is it just that we can get multiple times to `.Now()` that come back with the same wall time? That seems unlikely or at least rare. Are the logical ticks introduced when conflicts occur? If so, we can potentially eliminate a chunk of them on transaction retries. If a transaction retries, its new timestamp is a mandated minimum. But it could also use any larger timestamp <= `hlc.Now()`, which almost certainly includes one that has zero logical component. And even for doing that, we should understand if/when the complexity will have a measurable impact.

**Comment by petermattis** [comment]
This is super easy to reproduce so understanding where the ticks are coming from should be straightforward.

**Comment by tschottdorf** [comment]
Which experiment are you running to repro?

**Comment by petermattis** [comment]
I did nothing special other than instrument `hlc.Now` to print when the logical component was ticked and then run that binary on one of the test clusters under `kv` load. Specifically, I tested on `denim`.

**Comment by petermattis** [comment]
I took another look at this, instrumenting `hlc.Clock` to record the physical time at which we started incrementing the logical component and logging it when we reset it to 0. Over a 10s span, I recorded 5613 such events. The largest delta was 1.36ms. Only 3 events recorded ticking the logical component for longer than 1ms. The largest number of consecutive logical ticks was 89 and that happened over a 194us time period.
